src/dbnode/storage/throttler.go

```diff
@@ -0,0 +1,126 @@
+package storage
+
+import (
+	"sync"
+)
+
+// Throttler controls fair access to limited resources.
+type Throttler struct {
+	sync.Mutex
+
+	keyState map[string]*keyContext
+	// Each entry in the queue should be unique to avoid unfair access
+	keyQueue []string
+
+	globalCurrentWeight int
+	globalMaxWeight     int
+}
+
+type keyContext struct {
+	// Requests for weight to be granted which are waiting.
+	waiting []request
+	// Currently granted weight for this key.
+	currentWeight int
+}
+
+type request struct {
+	blockCh chan struct{}
+	weight  int
+}
+
+// Acquire blocks until the requested weight is granted to the specified key.
+func (t *Throttler) Acquire(key string, weight int) {
+	t.Lock()
+	defer t.Unlock()
+
+	maxWeightPerKey := t.maxWeightPerKey()
+
+	currentKey, alreadyExists := t.keyState[key]
+	if !alreadyExists {
+		t.keyState[key] = &keyContext{
+			currentWeight: 0,
+			waiting:       make([]request, 0),
+		}
+	}
+
+	// If new weight would keep the key under the per-key limit then grant.
+	// Otherwise enqueue and block.
+	if currentKey.currentWeight+weight < maxWeightPerKey {
+		currentKey.currentWeight += weight
+		t.globalCurrentWeight += weight
+	} else {
+		blockCh := make(chan struct{}, 0)
+		currentKey.waiting = append(currentKey.waiting, request{
+			blockCh: blockCh,
+			weight:  weight,
+		})
+
+		// If this is first request by the key, then enqueue it for releasing.
+		if !alreadyExists {
+			t.keyQueue = append(t.keyQueue, key)
+		}
+
+		// Block until enough weight is released.
+		<-blockCh
```

Comment:
This is going to deadlock, since you are blocking on the channel with the mutex locked.

Reply:
👀  good call

Reply:
Should be addressed now by locking within an internal func and returning the channel to block on if non-nil